File: lib/djs.rb

require 'thread'

module DJS
  TYPE_HANDSHAKE = 0
  TYPE_CONNECT   = 1
  TYPE_RESPONSE  = 2
  TYPE_MESSAGE   = 3
  TYPE_CLOSE     = 4
  TYPE_CALLBACK  = 5
  TYPE_RECONNECT = 6
  TYPE_RPC       = 7

  class DJSError < RuntimeError; end
  class DJSBadRequestError < DJSError; end

  def start(klass = DJSConnection, config = {})
    @server = DJSServer.new(klass, config)
    DJSConnections.server = @server
  end
  module_function :start

  def stop
    puts "STOP"
  end
  module_function :stop

  def log(log)
    puts Thread.current.to_s + " " + log.to_s
  end
  module_function :log

  def response(request)
    if request =~ /\A(\d)\x00(\d+)\x00(.*)\z/
      case $1.to_i
      when TYPE_HANDSHAKE
        response = @server.handshake
      when TYPE_CONNECT
        response = @server.connect($2.to_i)
      when TYPE_RESPONSE
        response = @server.response($2.to_i, $3)
      when TYPE_MESSAGE
      when TYPE_CLOSE
      when TYPE_CALLBACK
        response = @server.callback($2.to_i, $3)
      when TYPE_RECONNECT
        response = @server.reconnect($2.to_i)
      when TYPE_RPC
        response = @server.rpc($2.to_i)
      else

      end
    else
      Thread.new do
        onerror DJSBadRequestError.new
      end
      response = nil
    end

    return response.to_s
  end
  module_function :response

  def connections
    @server.djs_connections
  end
  module_function :connections

  def add_task(conn, name, *args)
    @server.add_task(conn, name, *args)
  end
  module_function :add_task

  class DJSConnections

    def self.server=(server)
      @@server = server
    end

    def initialize(connections)
      @connections = connections
    end

    def [](pattern)
      if pattern == "*"
        return self
      end
      DJSConnections.new(Hash.new)
    end

    def method_missing(name, *args)
      @connections.values.each { |conn|
        next unless conn.type == :main
        rpc_conn = @@server.create_rpc_connection(conn.cid, name, args)
        prx = DJSRpcProxyObject.new(rpc_conn)
        conn.add_proxy(prx, true)
      }
    end
  end

  class DJSServer
    TERMINAL = "\x00"
    RECONNECT = ""

    def initialize(klass = nil, config = {})
      @klass = klass
      @config = config

      @fiber = nil
      @connections = Hash.new
      @djs_connections = DJSConnections.new(@connections)
      @tasks = Queue.new
      @group = ThreadGroup.new
      @thread = run
    end
    attr_reader :connections, :djs_connections, :fiber

    def run
      Thread.start do
        begin
          while true
            main_loop
          end
        ensure
          # TODO kill thread
        end
      end
    end

    def main_loop
      Thread.start(@tasks.pop) do |task|
        conn = task[0]
        name = task[1]
        args = task[2]
        @fiber = Fiber.new do
          if args
            conn.send(name, *args)
          else
            conn.send(name)
          end
          conn.add_proxy(nil, true)
          TERMINAL
        end
        req = @fiber.resume
        if req == TERMINAL
          if conn.type == :main
            conn.request.push RECONNECT
          else
            @connections.delete(conn.__id__)
            conn.request.push TERMINAL
            return
          end
        else
          conn.request.push req
        end
        while rsp = conn.response.pop
          req = @fiber.resume(rsp)
          if req == TERMINAL
            if conn.type == :main
              conn.request.push RECONNECT
            else
              @connections.delete(conn.__id__)
              conn.request.push TERMINAL
              return
            end
          else
            conn.request.push req
          end
        end
      end
    end

    def handshake
      conn = @klass.new(:main)
      @connections[conn.__id__] = conn
      return conn.__id__.to_s
    end

    def connect(id)
      conn = @connections[id]
      add_task conn, :on_open
      req = conn.request.pop
      return req
    end

    def reconnect(id)
      conn = @connections[id]
      req = conn.request.pop
      return req
    end

    def response(id, rsp)
      conn = @connections[id]
      conn.response.push rsp
      req = conn.request.pop
      return req
    end

    def callback(cid, method)
      conn = @klass.new(:callback, cid)
      @connections[conn.__id__] = conn
      add_task conn, :on_callback, method
      req = conn.request.pop
      return conn.__id__.to_s + "\x00" + req
    end

    def rpc(id)
      conn = @connections[id]
      add_task conn, conn.rpc[0], conn.rpc[1]
      req = conn.request.pop
      return req
    end

    def add_task(conn, name, *args)
      @tasks.push [conn, name, args]
    end

    def create_rpc_connection(cid, name, args)
      conn = @klass.new(:rpc, cid)
      @connections[conn.__id__] = conn
      conn.set_rpc(name, args)
      return conn
    end
  end

  class DJSConnection
    SEP = "\x00"

    def initialize(type, cid = 0)
      @type = type
      if @type == :main
        @cid = self.__id__
      else
        @cid = cid
      end
      @request = Queue.new
      @response = Queue.new
      @proxies = Hash.new
      @proxy_ids = Array.new
      @mutex = Mutex.new
    end

    attr_accessor :type, :request, :response, :proxies
    attr_reader :cid

    def set_rpc(name, args)
      @rpc = [name, args]
    end

    def rpc
      @rpc
    end

    def add_proxy(proxy, flush_now = false)
      @mutex.synchronize {
        if proxy
          @proxy_ids << proxy.__id__
          @proxies[proxy.__id__] = proxy
        end
        # TODO length
        if flush_now || @proxy_ids.length == 100
          flush
        end
      }
    end

    def flush
      if @proxy_ids.empty?
        return
      end

      json = "0\x00"
      @proxy_ids.each { |id|
        json << @proxies[id].__json << SEP
      }
      json[-1] = ""
      rsp = Fiber.yield(json)
      while rsp != '{}'
        info = eval(rsp)

        if info.key?("type")
          result = info["type"].send(info["content"], *info["args"])
          proxy = @proxies[info["id"]]
          proxy = @proxies[proxy.info[:id]]
          proxy.origin = result
          proxy.solved = true

          json = "1\x00{\"id\":" + DJS.to_json(info["id"]) + ",\"origin\":" + DJS.to_json(result) + "}"
          rsp = Fiber.yield(json)
        else
          info.each { |key, value|
            if @proxies.key?(key)
              @proxies[key].origin = value
              @proxies[key].solved = true
            end
          }
          rsp = "{}"
        end
      end

      @proxies.clear
      @proxy_ids.clear
    end

    def on_callback(method)
      self.method(method).call
    end

    # Event handler
    def on_open

    end

    def on_close

    end

    def on_message(msg)

    end

    def on_error(err)

    end

    # Root Javascript Object
    def window
      DJSProxyObject.new(self, {:type => :window})
    end

    def document
      DJSProxyObject.new(self, {:type => :document})
    end

    def frame
      DJSProxyObject.new(self, {:type => :Frame})
    end

    def history
      DJSProxyObject.new(self, {:type => :history})
    end

    def location
      DJSProxyObject.new(self, {:type => :location})
    end
  end

  class DJSProxyObject
    def initialize(conn, info)
      @conn = conn
      @info = info
      @origin = nil
      @solved = false
      @info[:id] = __id__
      @info[:cid] = conn.cid
    end

    attr_accessor :origin, :solved, :info

    def method_missing(name, *args, &block)
      if @solved
        return @origin.send(name, *args, &block)
      end

      if block
        @conn.add_proxy(nil, true)
        return @origin.send(name, *args, &block)
      end

      if name.to_s =~ /^.*=$/ && args && Symbol === args[0]
        name = "{}" + name.to_s[/[^=]+/]
      end

      proxy = DJSProxyObject.new(@conn, {:type => self, :content => name, :args => args})
      @conn.add_proxy proxy
      return proxy
    end

    def sync
      return @origin if @solved
      @conn.add_proxy(nil, true)
      return @solved ? @origin : self
    end

    def __json
      DJS.to_json @info
    end

    def __to_s
      #TODO refs
      if DJSProxyObject === @info[:type]
        return "REFS[" + __id__.to_s + "].origin"
      else
        return @info[:type].to_s
      end
    end
  end

  class DJSRpcProxyObject < DJSProxyObject
    def initialize(conn)
      super(conn, {:type => 'rpc', :content => conn.__id__})
    end
  end

  def to_json(obj)
    case obj
    when Hash
      json = '{'
      obj.each { |key, value|
        json << '"' << key.to_s << '":' << to_json(value) << ','
      }
      if (json.length > 1)
        json[-1] = '}'
      else
        json << '}'
      end
      return json
    when Array
      json = '['
      obj.each { |item|
        json << to_json(item) << ','
      }
      if (json.length > 1)
        json[-1] = ']'
      else
        json << ']'
      end
      return json
    when DJSProxyObject
      return obj.__to_s
    when Numeric
      return obj.to_s
    when String
      return '"' + obj.gsub('"', '\\"') + '"'
    when TrueClass, FalseClass
      return obj.to_s
    else
      return '"' + obj.to_s + '"'
    end
  end
  module_function :to_json
end

File: lib/djs.js

var WS;
var REFS;
var SERVER;
var CALLBACKS;
var handshaked;
var isWebsocket;
var TASKS;

// WebSocket Implementation
function djs_start_websocket(server) {
    WS = new WebSocket(server);
    WS.onopen = on_open;
    WS.onclose = on_close;
    WS.onmessage = on_message;
    WS.onerror = on_error;
    handshaked = false;
    REFS = new Array();
    TASKS = new Array();
    CALLBACKS = new Array();
    isWebsocket = true;
}

function websocket_handshake() {
    WS.send("0\x000\x00");
}

function on_open() {
    websocket_handshake();
    console.log("opened");
}

function on_close() {
    console.log("closed");
}

function on_message(event) {
    if (!handshaked) {
        handshaked = true;
        DJS_ID = event.data;
        WS.send("1\x00" + DJS_ID + "\x00");
    } else {
        if (event.data == "\x00") {
            return;
        }
        var rsp = djs_eval(event.data);
        WS.send("2\x00" + DJS_ID + "\x00" + rsp);
    }
}

function on_error(error) {
    console.log("error");
}

function oncallback(cid, method, event) {
    WS.send("5\x00" + cid + "\x00" + method);
}

// AJAX Implementation
function djs_start_ajax(server) {
    SERVER = server;
    REFS = new Array();
    TASKS = new Array();
    CALLBACKS = new Array();
    isWebsocket = false;

    handshake();
}

function djs_send(msg, callback) {
    message(msg, callback);
}

function getXmlHttpRequestObject() {
    var xhr;
    if (XMLHttpRequest) {
        xhr = new XMLHttpRequest();
    } else {
        try {
            xhr = new ActiveXObject('MSXML2.XMLHTTP.6.0');
        } catch (e) {
            try {
                xhr = new ActiveXObject('MSXML2.XMLHTTP.3.0');
            } catch (e) {
                try {
                    xhr = new ActiveXObject('MSXML2.XMLHTTP');
                } catch (e) {
                    xhr = null;
                    alert("This browser does not support XMLHttpRequest.");
                }
            }
        }
    }

    return xhr;
}

function handshake() {
    var xhr = getXmlHttpRequestObject();
    xhr.open("POST", SERVER);
    xhr.setRequestHeader("Content-Type", "text/plain");
    xhr.onreadystatechange = function() {
        if (xhr.readyState === 4 && xhr.status === 200) {
            if (xhr.responseText == "\x00") {
                return;
            }
            var id = xhr.responseText;
            connect(id);
        }
    };
    xhr.send("0\x000\x00");
}

function connect(id) {
    var xhr = getXmlHttpRequestObject();
    xhr.open("POST", SERVER);
    xhr.setRequestHeader("Content-Type", "text/plain");
    xhr.onreadystatechange = function() {
        if (xhr.readyState === 4 && xhr.status === 200) {
            var rsp = xhr.responseText;
            if (rsp == "\x00") {
                return;
            }
            if (rsp == "") {
                reconnect(id);
                return;
            }
            var req = djs_eval(rsp);
            response(id, req);
        }
    };
    xhr.send("1\x00" + id + "\x00");
}

function reconnect(id) {
    var xhr = getXmlHttpRequestObject();
    xhr.open("POST", SERVER);
    xhr.setRequestHeader("Content-Type", "text/plain");
    xhr.onreadystatechange = function() {
        if (xhr.readyState === 4 && xhr.status === 200) {
            var rsp = xhr.responseText;
            if (rsp == "\x00") {
                return;
            }
            if (rsp == "") {
                reconnect(id);
                return;
            }
            var req = djs_eval(rsp);
            response(id, req);
        }
    };
    xhr.send("6\x00" + id + "\x00");
}

function response(id, rsp) {
    var xhr = getXmlHttpRequestObject();
    xhr.open("POST", SERVER);
    xhr.setRequestHeader("Content-Type", "text/plain");
    xhr.onreadystatechange = function() {
        if (xhr.readyState === 4 && xhr.status === 200) {
            var rsp = xhr.responseText;
            if (rsp == "\x00") {
                return;
            }
            if (rsp == "") {
                reconnect(id);
                return;
            }
            var req = djs_eval(rsp);
            response(id, req);
        }
    };
    xhr.send("2\x00" + id + "\x00" + rsp);
}

function callback(cid, method, event) {
    var xhr = getXmlHttpRequestObject();
    xhr.open("POST", SERVER);
    xhr.setRequestHeader("Content-Type", "text/plain");
    xhr.onreadystatechange = function() {
        if (xhr.readyState === 4 && xhr.status === 200) {
            var rsp = xhr.responseText;
            var id = rsp.substring(0, rsp.indexOf("\x00"));
            if (rsp == "\x00") {
                return;
            }
            var req = djs_eval(rsp.substring(rsp.indexOf("\x00") + 1));
            response(id, req);
        }
    }
    xhr.send("5\x00" + cid + "\x00" + method);
}

function rpc(id) {
    var xhr = getXmlHttpRequestObject();
    xhr.open("POST", SERVER);
    xhr.setRequestHeader("Content-Type", "text/plain");
    xhr.onreadystatechange = function() {
        if (xhr.readyState === 4 && xhr.status === 200) {
            var rsp = xhr.responseText;
            if (rsp == "\x00") {
                return;
            }
            var req = djs_eval(rsp);
            response(id, req);
        }
    }
    xhr.send("7\x00" + id + "\x00");
}

function djs_eval(json) {
    var reqs = json.split("\x00");
    var info;
    var id;
    var cid;
    var obj;
    var content;
    var args;
    var result;
    var i;
    if (reqs[0] == "1") {
        for (i = 1; i < reqs.length; i++) {
            info = eval("(" + reqs[i] + ")");
            REFS[info.id].origin = info.origin;
        }
    } else {
        for (i = 1; i < reqs.length; i++) {
            TASKS.push(reqs[i]);
        }
    }

    var req;
    var rsp = "{";
    while (TASKS.length > 0) {
        req = TASKS.shift();
        info = eval("(" + req + ")");
        id = info.id;
        cid = info.cid;
        obj = info.type;
        content = info.content;
        args = info.args;

        if (obj == "rpc") {
            rpc(content);
            // TODO
            continue;
        }

        if (info.content.indexOf("{}") == 0) {
            obj[info.content.substring(2)] = createCallback(cid, info.args[0]);
            result = null;
        } else {
            // TODO obj[content]
            if (isPrimitive(obj) && !obj[content]) {
                to_ruby_object(info.id, null);
                info.type = obj;
                return toRubyHash(info);
            }

            if (args && args.length > 0) {
                if (content.match(/^.+=$/)) {
                    obj[content.substring(0, content.length - 1)] = args[0];
                    result = null;
                } else {
                    // TODO
                    if (args.length > 1) {
                        result = obj[content].apply(obj, args);
                    } else {
                        result = obj[content](args[0]);
                    }
                }
            } else {
                if (typeof obj[content] == "function") {
                    result = obj[content]();
                } else {
                    result = obj[content];
                }
            }
        }

        obj = to_ruby_object(id, result);
        if (obj) {
            rsp += id + "=>" + obj + ",";
        }
    }

    if (rsp.length > 1) {
        rsp = rsp.substring(0, rsp.length - 1) + "}";
    } else {
        rsp += "}";
    }
    return rsp;
}

function createRefObj(id, obj) {
    var refObj = new RefObj(id, obj);
    REFS[id] = refObj;
    return refObj;
}

function createCallback(cid, method) {
    return function(cid, method) {
        return function(e) {
            if (isWebsocket) {
                oncallback(cid, method, e);
            } else {
                callback(cid, method, e);
            }
        }
    }(cid, method);
}

function RefObj(id, obj) {
    this.id = id;
    this.origin = obj;
}

function to_ruby_object(id, obj) {
    if (obj == undefined || obj == null) {
        return null;
    }

    // TODO all type?
    var type = typeof obj;
    if (type == "string") {
        return "'" + obj.replace(/'/g, "\\'") + "'";
    } else if (type == "number" || type == "boolean") {
        return obj;
    } else {
        createRefObj(id, obj);
        return null;
    }
}

function isPrimitive(obj) {
    var type = typeof obj;
    return type == "string" || type == "number" || type == "boolean";
}

function toRubyHash(hash) {
    var ret = '{';
    for (key in hash) {
        ret += '"' + key + '"=>';
        if (typeof hash[key] == "string") {
            ret += '"' + hash[key] + '",';
        } else {
            // TODO
            if (hash[key] instanceof Array && hash[key].length == 0) {
                ret += '[],';
            } else {
                ret += hash[key] + ',';
            }
        }
    }
    if (ret.length > 1) {
        ret = ret.substring(0, ret.length - 1);
    }
    ret += '}';
    return ret;
}
